packages/chain-adapters/src/utxo/UtxoBaseAdapter.ts

```diff
@@ -228,39 +228,30 @@ export abstract class UtxoBaseAdapter<T extends UtxoChainId> implements IChainAd
 
       const bip44Params = this.getBIP44Params({ accountNumber, accountType, isChange, index })
 
-      if (pubKey) {
-        const account = await this.getAccount(pubKey)
-        const index = bip44Params.isChange
-          ? account.chainSpecific.nextChangeAddressIndex
-          : account.chainSpecific.nextReceiveAddressIndex
-
-        if (index === undefined)
-          throw new Error(`UtxoBaseAdapter: Could not fetch address index from unchained`)
-        const address = account.chainSpecific.addresses?.[index]?.pubkey
-        if (!address)
-          throw new Error(
-            `UtxoBaseAdapter: Could not fetch address from unchained at index ${index}`,
+      const account = await (async () => {
+        if (pubKey || bip44Params.index === undefined) {
+          return this.getAccount(
+            pubKey ?? (await this.getPublicKey(wallet, accountNumber, accountType)).xpub,
           )
+        }
+      })()
 
-        return address
-      }
+      const nextIndex = bip44Params.isChange
+        ? account?.chainSpecific.nextChangeAddressIndex
+        : account?.chainSpecific.nextReceiveAddressIndex
 
-      const getNextIndex = async () => {
-        const { xpub } = await this.getPublicKey(wallet, accountNumber, accountType)
-        const account = await this.getAccount(xpub)
+      const indexOrNextIndex = bip44Params.index ?? nextIndex ?? 0
 
-        return bip44Params.isChange
-          ? account.chainSpecific.nextChangeAddressIndex
-          : account.chainSpecific.nextReceiveAddressIndex
-      }
+      const address = await (() => {
+        if (pubKey) return account?.chainSpecific.addresses?.[indexOrNextIndex]?.pubkey
```

Comment:
note, it doesn't look like we call `getAddress` with `isChange: true` anywhere, but this won't take into consideration receive vs change address in conjunction with the supplied index.

also, if when it comes to custom derivation paths which I saw a request for, we may have to rethink this or keep track of gap limit so we don't try to request an index that isn't derived by blockbook or worse, use an index that actually falls into the change address section of the addresses array.

Reply:
I think both of these are mitigated by the fact this is used exclusively for your account 0 for THORFi operations

Reply:
correct, as it stands, the logic works, but we will need to be cognizant if/when supporting custom address index or if ever using this in conjunction with `isChange: true`.